src/Symfony/Bundle/WebServerBundle/DependencyInjection/WebServerExtension.php

```diff
@@ -27,8 +27,31 @@ public function load(array $configs, ContainerBuilder $container)
         $loader = new XmlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
         $loader->load('webserver.xml');
 
+        $publicDirectory = $this->getPublicDirectory($container);
+        $container->getDefinition('web_server.command.server_run')->replaceArgument(0, $publicDirectory);
+        $container->getDefinition('web_server.command.server_start')->replaceArgument(0, $publicDirectory);
+
         if (!class_exists(ConsoleFormatter::class)) {
             $container->removeDefinition('web_server.command.server_log');
         }
     }
+
+    private function getPublicDirectory(ContainerBuilder $container)
+    {
+        $kernelProjectDir = $container->getParameter('kernel.project_dir');
+        $publicDir = 'public';
+        $composerFilePath = $kernelProjectDir.'/composer.json';
+
+        if (!file_exists($composerFilePath)) {
+            return $kernelProjectDir.'/'.$publicDir;
+        }
+
+        $composerConfig = json_decode(file_get_contents($composerFilePath), true);
+
+        if (isset($composerConfig['extra']['public-dir'])) {
+            $publicDir = $composerConfig['extra']['public-dir'];
```

Comment:
Should we do a `ltrim($publicDir, '/')` here to be completely sure that we won't have double slashes later?

Reply:
why should we care?